ambry-frontend/src/main/java/com/github/ambry/frontend/NamedBlobDeleteHandler.java

```diff
@@ -0,0 +1,109 @@
+import com.github.ambry.named.NamedBlobRecord;
+import com.github.ambry.rest.RequestPath;
+import com.github.ambry.rest.RestRequest;
+import com.github.ambry.rest.RestResponseChannel;
+import com.github.ambry.rest.RestServiceException;
+import com.github.ambry.utils.Utils;
+import java.util.Collections;
+
+import static com.github.ambry.rest.RestUtils.*;
+
+
+/**
+ * {@link NamedBlobDeleteHandler} is a simple wrapper around {@link DeleteBlobHandler} as the only difference between
+ * named blob delete and regular blob delete is that named blob provides a named blob id instead of a regular blob id.
+ *
+ * In the {@link #handle} method, we would first replace the named blob id in the {@link RequestPath} with the
+ * corresponding regular blob id and then call {@link DeleteBlobHandler#handle} to handle the request.
+ */
+class NamedBlobDeleteHandler {
+  private final AccountAndContainerInjector accountAndContainerInjector;
+  private final FrontendMetrics metrics;
+  private final DeleteBlobHandler deleteBlobHandler;
+  private final NamedBlobDb namedBlobDb;
+
+  /**
+   * Construct a {@link NamedBlobDeleteHandler} for handling requests for deleting blobs in named blob accounts.
+   * @param accountAndContainerInjector The {@link AccountAndContainerInjector} to use.
+   * @param namedBlobDb The {@link NamedBlobDb} to use.
+   * @param metrics The {@link FrontendMetrics} instance where metrics are recorded
+   * @param deleteBlobHandler The {@link DeleteBlobHandler}.
+   */
+  NamedBlobDeleteHandler(AccountAndContainerInjector accountAndContainerInjector, NamedBlobDb namedBlobDb,
+      FrontendMetrics metrics, DeleteBlobHandler deleteBlobHandler) {
+    this.accountAndContainerInjector = accountAndContainerInjector;
+    this.namedBlobDb = namedBlobDb;
+    this.metrics = metrics;
+    this.deleteBlobHandler = deleteBlobHandler;
+  }
+
+  /**
+   * Asynchronously handle named blob delete requests.
+   * @param restRequest the {@link RestRequest} that contains the request parameters and body.
+   * @param restResponseChannel the {@link RestResponseChannel} where headers should be set.
+   * @param callback the {@link Callback} to invoke when the response is ready (or if there is an exception).
+   * @throws RestServiceException
+   */
+  void handle(RestRequest restRequest, RestResponseChannel restResponseChannel, Callback<Void> callback)
+      throws RestServiceException {
+    RequestPath requestPath = getRequestPath(restRequest);
+    // Get blob id and reconstruct RequestPath
+    // this is the three-part named blob id, including account name, container name and the custom blob name.
+    String namedBlobId = requestPath.getOperationOrBlobId(false);
+    NamedBlobPath namedBlobPath = NamedBlobPath.parse(namedBlobId, Collections.emptyMap());
+    // This is a three steps operation
+    // 1. Get regular blob id from named blob id.
+    // 2. Delete regular blob id from Router
+    // 3. Delete database row.
```

Comment:
@SophieGuo410 can correct me. Seems like currently it is 2 step process, we delete (soft delete) the Named blob ID first and then delete the regular Blob ID. In case step 2 fails, it seems like the blob could be orphaned (Not sure though if this is actually the case. May be we need to implement to delete orphaned blobs asynchronously)?

If we make it as 3 step process to GET blob ID first from NamedBlobDB, delete the regular blob ID from Ambry and delete DB row, seems like we won't need need asynchronous orphaned blob clean up. But, would it increase the DELETE latency due to 2 network calls to DB?

Reply:
Once Mike's update named blob feature is done, I think we are able to handle it in a better way. We will memorize the old blob in namedBlobDB until it gets purged by background deleter. 
One thing I want to point out is, if step3 failed, we will return fail the customer, which means if delete failed, user should still be able to get the blob. But in this sequence, the real blob will be gone, which I think will bring trouble.

Reply:
the contract of delete is not like this. In regular blobs, once the client issues a delete, it's considered as deleted, whether ambry returns 202 or 503. We should use the same logic here.